lib/src/nogoods.rs

```diff
@@ -0,0 +1,785 @@
+pub struct NoGood {
+    active: RoaringBitmap,
+    value: RoaringBitmap,
+}
+
+impl Eq for NoGood {}
+impl PartialEq for NoGood {
+    fn eq(&self, other: &Self) -> bool {
+        self.active
+            .borrow()
+            .bitxor(other.active.borrow())
+            .is_empty()
+            && self.value.borrow().bitxor(other.value.borrow()).is_empty()
+    }
+}
+
+impl NoGood {
+    /// Creates a [NoGood] from a given Vector of [Terms][Term].
+    pub fn from_term_vec(term_vec: &[Term]) -> NoGood {
+        let mut result = Self::default();
+        term_vec.iter().enumerate().for_each(|(idx, val)| {
+	    let idx:u32 = idx.try_into().expect("no-good learner implementation is based on the assumption that only u32::MAX-many variables are in place");
+            if val.is_truth_value() {
+                result.active.insert(idx);
+                if val.is_true() {
+                    result.value.insert(idx);
+                }
+            }
+        });
+        result
+    }
+
+    /// Creates a [NoGood] representing an atomic assignment.
+    pub fn new_single_nogood(pos: usize, val: bool) -> NoGood {
+        let mut result = Self::default();
+        let pos:u32 = pos.try_into().expect("nog-good learner implementation is based on the assumption that only u32::MAX-many variables are in place");
+        result.active.insert(pos);
+        if val {
+            result.value.insert(pos);
+        }
+        result
+    }
+
+    /// Returns [None] if the pair contains inconsistent pairs.
+    /// Otherwise it returns a [NoGood] which represents the set values.
+    pub fn try_from_pair_iter(
+        pair_iter: &mut impl Iterator<Item = (usize, bool)>,
+    ) -> Option<NoGood> {
+        let mut result = Self::default();
+        let mut visit = false;
+        for (idx, val) in pair_iter {
+            visit = true;
+            let idx:u32 = idx.try_into().expect("no-good learner implementation is based on the assumption that only u32::MAX-many variables are in place");
+            let is_new = result.active.insert(idx);
+            let upd = if val {
+                result.value.insert(idx)
+            } else {
+                result.value.remove(idx)
+            };
+            // if the state is not new and the value is changed
+            if !is_new && upd {
+                return None;
+            }
+        }
+        if visit {
+            Some(result)
+        } else {
+            None
+        }
+    }
+
+    /// Creates an updated [Vec<Term>], based on the given [&[Term]] and the [NoGood].
+    /// The parameter _update_ is set to [`true`] if there has been an update and to [`false`] otherwise
+    pub fn update_term_vec(&self, term_vec: &[Term], update: &mut bool) -> Vec<Term> {
+        *update = false;
+        term_vec.iter().enumerate().map(|(idx,val)|{
+	    let idx:u32 = idx.try_into().expect("no-good learner implementation is based on the assumption that only u32::MAX-many variables are in place");
```

Comment:
Break the string literal to work around rust-lang/rustfmt#3863, which prevents formatting of the remainder of the function.

```suggestion
                let idx: u32 = idx.try_into().expect(
                    "no-good learner implementation is based on the assumption \
                     that only u32::MAX-many variables are in place",
                );
```